src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1401 @@
+
+      auto& referrerUrl = registry.lookup(js, referrer).map(
+          [&](IsolateModuleRegistry::Entry& entry) -> const Url& {
+        switch (entry.module.type()) {
+          case Module::Type::BUNDLE: {
+            type = ResolveContext::Type::BUNDLE;
+            break;
+          }
+          case Module::Type::BUILTIN: {
+            type = ResolveContext::Type::BUILTIN;
+            break;
+          }
+          case Module::Type::BUILTIN_ONLY: {
+            type = ResolveContext::Type::BUILTIN_ONLY;
+            break;
+          }
+          case Module::Type::FALLBACK: {
+            type = ResolveContext::Type::BUNDLE;
+            break;
+          }
+        }
+        return entry.specifier.specifier;
+      }).orDefault(ModuleBundle::BundleBuilder::BASE);
+
+      KJ_IF_SOME(url, referrerUrl.tryResolve(spec)) {
+        // Make sure that percent-encoding in the path is normalized so we can match correctly.
+        auto normalized = url.clone(Url::EquivalenceOption::NORMALIZE_PATH);
+        ResolveContext resolveContext = {
+          .type = type,
+          .source = ResolveContext::Source::STATIC_IMPORT,
+          .specifier = normalized,
+          .referrer = referrerUrl,
+          .rawSpecifier = spec.asPtr(),
+        };
+        // TODO(soon): Add import assertions to the context.
+
+        return registry.resolve(js, resolveContext);
+      }
+
+      js.throwException(js.error(kj::str("Invalid module specifier: "_kj, specifier)));
+    }, [&](Value exception) -> v8::MaybeLocal<v8::Module> {
+      // If there are any synchronously thrown exceptions, we want to catch them
+      // here and convert them into a rejected promise. The only exception are
+      // fatal cases where the isolate is terminating which won't make it here
+      // anyway.
+      js.v8Isolate->ThrowException(exception.getHandle(js));
+      return v8::MaybeLocal<v8::Module>();
+    });
+  } catch (...) {
+    kj::throwFatalException(kj::getCaughtExceptionAsKj());
+  }
+}
+
+// The fallback module bundle calls a single resolve callback to resolve all modules
+// it is asked to resolve. Instances must be thread-safe.
+class FallbackModuleBundle final : public ModuleBundle {
+public:
+  FallbackModuleBundle(Builder::ResolveCallback&& callback)
+      : ModuleBundle(Type::FALLBACK),
+        callback(kj::mv(callback)) {}
+
+  kj::Maybe<Module&> resolve(const ResolveContext& context) override {
+    {
+      auto lock = cache.lockShared();
+      KJ_IF_SOME(found, lock->storage.find(context.specifier)) {
+        return kj::Maybe<Module&>(const_cast<Module&>(*found));
+      }
+      KJ_IF_SOME(found, lock->aliases.find(context.specifier)) {
+        return kj::Maybe<Module&>(*found);
+      }
+    }
+
+    {
+      auto lock = cache.lockExclusive();
+      KJ_IF_SOME(resolved, callback(context)) {
+        Module& module = *resolved;
+        lock->storage.upsert(context.specifier.clone(), kj::mv(resolved));
```

Comment:
Conceptual question: Is the clone() call needed for memory safety here or is it just based on how URL works? I noticed that Url::clone takes an option parameter that influences cloning, but it is not being used here.

Reply:
A bit of both. The lifetimes of these things are a bit different so cloning is the safest option.